message_view: Add unread divider in message list.

vrongmeal opened the pull request:
This adds a divider for the first unread message of the narrow. It
appears only when the narrow is a topic or pm with a person/group.
If the message includes this divider it also includes the sender;
If the message has a date divider, it replaces it with a custom
version of the date divider which include the new messages text with
the dates above and below, else the divider just has the new messages
text. The divider does not appear when the message is first in it's
group.

Fixes #12518 

**GIFs or Screenshots:**

**With date divider**
![Screenshot 2019-06-18 at 12 51 22 AM](https://user-images.githubusercontent.com/33068691/59633560-5c24fe80-916a-11e9-937c-4407db251d2b.png)

**Without date divider**
![Screenshot 2019-06-18 at 12 51 42 AM](https://user-images.githubusercontent.com/33068691/59633568-60e9b280-916a-11e9-91bb-2b921c7e4054.png)

**Night mode**
![Screenshot 2019-06-18 at 12 52 07 AM](https://user-images.githubusercontent.com/33068691/59633591-75c64600-916a-11e9-96fe-929d21c0cf33.png)

vrongmeal committed:
message_list_view: Add unread divider in message list.

This adds a divider for the first unread message of the narrow. It
appears only when the narrow is a topic or pm with a person/group.
If the message includes this divider it also includes the sender;
If the message has a date divider, it replaces it with a custom
version of the date divider which include the new messages text with
the dates above and below, else the divider just has the new messages
text. The divider does not appear when the message is first in it's
group.

So as to not extend the unread divider, the unread marker is made
smaller.

Fixes #12518

timabbott commented:
It took me a while to reproduce this working, because I wasn't looking at a topic narrow; FYI @rishig for when you're testing this, it only appears in those.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -132,6 +132,59 @@ function update_message_date_divider(opts) {
+                                  unread_divider_html.html() +
+                                  '</span>';
+    } else {
+        msg.unread_divider_html = '<span>' + NEW_MESSAGES_TEXT + '</span>';
```
Generally, building up HTML with `+` can be highly problematic for security.  It'd be better to do `$("span").text(NEW_MESSAGES_TEXT)` if we're constructing this with jQuery; but it might be even better to just use a very small handlebars template for this and pass in the details via the context.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -132,6 +132,59 @@ function update_message_date_divider(opts) {
+    if (!(is_topic || is_pm)) {
+        clear_unread_messages_divider(curr_msg_container);
+        return;
+    }
```
I think it might make sense to put this `is_topic` check as the first thing in the function?  Because in that case, it doesn't matter what the details of this particular message are, so it's sorta a higher-level check.

timabbott commented on static/js/message_list_view.js:
```diff
@@ -132,6 +132,59 @@ function update_message_date_divider(opts) {
+    var is_topic = filter.has_operator('stream') && filter.has_operator('topic');
+    var is_pm = filter.has_operator('pm-with') || filter.has_operator('group-pm-with');
+
+    if (!(is_topic || is_pm)) {
```
Let's add a comment before this check:
```
// Our unread message dividers, as currently conceived, don't really make sense in interleaved narrows,
// since we could have read messages after unread messages in those contexts.
// As a result, we currently don't offer these "New messages" notices in those views.
```

timabbott replied:
(Just for context, I expect we'll later add support for this in a wider range of views, and e.g. just have the text change to something more subtle that communicates "New messages, including some messages you read via other views").

timabbott commented on static/js/message_list_view.js:
```diff
@@ -132,6 +132,59 @@ function update_message_date_divider(opts) {
+    var NEW_MESSAGES_TEXT = 'NEW MESSAGES';
+
+    if (msg.want_date_divider) {
+        msg.want_date_divider = false;
```
I'm a little skeptical of settings `want_date_divider` to false; do we need to do that?  I could imagine some future code trying to check if there's a date divider being confused by this.

vrongmeal replied:
Yeah I removed it. The `want_unread_divider` takes preference in the template so it won't be a problem.

rishig commented:
cool, thanks @vrongmeal! Would it be possible to make the unread marker not extend into the "new messages" divider?

timabbott commented:
I posted a batch of comments.  I also have a higher-level question -- it appears that we need code related to this `unread_msgs` divider in every place where we had code for the `date_divider`.  Would it make sense to instead have both of these fit within a single "date_divider" concept that just has an additional `new_messages` flag on it, and just have this be an additional case that generates a date divider?  Might avoid a lot of duplication of the existing code for date dividers, which will get even bigger when we add support for these in between recipient rows.  Can you prototype whether that implementation strategy makes sense?

The other thing I'm worried about is this correctly handling what are intended to be no-op rerenders of the message list (e.g. when a message is edited around the point where the unread messages boundary is).  I think the right way to fix this will involve keeping track of which messages were unread "at the time the current narrow happened" (or on the `message_list` data structure), not in the `message_list_view` data structure, which is designed to be rebuilt from scratch.  Let's try to get the basic thing working before digging deep into this.

vrongmeal commented:
> The other thing I'm worried about is this correctly handling what are intended to be no-op rerenders of the message list (e.g. when a message is edited around the point where the unread messages boundary is). I think the right way to fix this will involve keeping track of which messages were unread "at the time the current narrow happened" (or on the message_list data structure), not in the message_list_view data structure, which is designed to be rebuilt from scratch. Let's try to get the basic thing working before digging deep into this.

I didn't really understand this point.

Are you saying that if the message is edited near the unread mark, the mark might shift? Because the divider is not updated during the re-rendering of the message so it won't be an issue.

vrongmeal commented:
> I also have a higher-level question -- it appears that we need code related to this unread_msgs divider in every place where we had code for the date_divider. Would it make sense to instead have both of these fit within a single "date_divider" concept that just has an additional new_messages flag on it, and just have this be an additional case that generates a date divider? Might avoid a lot of duplication of the existing code for date dividers, which will get even bigger when we add support for these in between recipient rows. Can you prototype whether that implementation strategy makes sense?

I don't really see much duplication. I agree that the `message_divider` code can be merged (made this change) but the only duplication in logic is checking if the `prev_message_container` is `undefined` or not. I guess I can move that inside the general `update_message_divider` function itself.

vrongmeal committed:
message_list_view: Extract update and clear message divider functions.

The `update_message_divider` and `clear_message_divider` update and
clear the date divider and the unread divider. Since both the dividers
have their update and clear functions called together, these functions
could be extracted easily.

vrongmeal committed:
message_list_view: Have one divider update and eliminate divider html.

This merges both the date and unread divider functions and elimates
the need for date_divider_html or unread_divider_html by adding a
handlebar template which renders using the divider_properties.

Also added message_divider.handlebars to custom_check list for
linting since it includes `style` attribute.

vrongmeal commented:
> cool, thanks @vrongmeal! Would it be possible to make the unread marker not extend into the "new messages" divider?

I think this started to happen when the date divider was inserted in-between the message container. I'll look into it :)

vrongmeal commented:
@timabbott I've made some changes here. Just want to get these approved first and then move on to shifting to handlebars for divider html (since it will require changes in date divider html too)...

timabbott commented:
> I didn't really understand this point.

Basically, if one does `message_list.view.rerender_preserving_scrolltop` after reading the new messages, my guess is the unread divider will disappear or move, because the `unread.js` state will have changed.  (We call that function in various cases when messages are edited, but you can just use it directly from the JS console for testing)

timabbott commented on static/js/message_list_view.js:
```diff
@@ -132,6 +132,85 @@ function update_message_date_divider(opts) {
+                                      curr_msg_container.want_unread_divider;
+    // Unread divider html is preferred over the date divider html
+    curr_msg_container.divider_html = curr_msg_container.unread_divider_html ||
+                                      curr_msg_container.date_divider_html;
```
I think this structure is an improvement, but I think we can go further: I think it should be possible to merge the `update_message_date_divider` and `update_unread_messages_divider` functions (basically just compute once whether the dates and different and whether this is the unread boundary, and if so, just do a `templates.render` on a single shared HTML template that renders dividers, with conditionals to handle the differences between the two very similar divider styles).  Does that make sense?

vrongmeal replied:
Yeah! I'll make this change.

vrongmeal commented:
> Basically, if one does message_list.view.rerender_preserving_scrolltop after reading the new messages, my guess is the unread divider will disappear or move, because the unread.js state will have changed. (We call that function in various cases when messages are edited, but you can just use it directly from the JS console for testing)

Currently, neither of the dividers update on re-rendering the message. Yes, this might be a problem for when the message moves into another group but yeah this will come later on.

vrongmeal commented:
@timabbott Pushed some changes here.

vrongmeal commented:
Hey @timabbott I rebased this with the latest master. Needs your review.
